allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,824 @@
+                                                             text_mask,
+                                                             span_starts,
+                                                             span_ends)
+        # Compute mention scores.
+        # Shape: (batch_size, num_spans, 1)
+        mention_scores = self._mention_scorer(self._mention_feedforward(span_embeddings))
+        mention_scores += span_mask.log()
+
+        # Prune based on mention scores.
+        num_spans_to_keep = int(math.floor(self._spans_per_word * text_embeddings.size(1)))
+
+        # Shape: (batch_size, num_spans_to_keep)
+        # These are indices (with values between 0 and num_spans) into
+        # the span_embeddings tensor.
+        top_span_indices = self._prune_and_sort_spans(mention_scores, num_spans_to_keep)
+
+        # Shape: (batch_size * num_spans_to_keep)
+        # torch.index_select only accepts 1D indices, but here
+        # we need to do select spans for each element in the batch.
+        # This reformats the indices to take into account their index into the batch.
+        flat_top_span_indices = util.flatten_and_batch_shift_indices(top_span_indices, span_starts.size(1))
+
+        # Select the span embeddings corresponding to the
+        # top spans based on the mention scorer.
+        # Shape: (batch_size, num_spans_to_keep, embedding_size)
+        top_span_embeddings = util.batched_index_select(span_embeddings,
+                                                        top_span_indices,
+                                                        flat_top_span_indices)
+        # Shape: (batch_size, num_spans_to_keep, 1)
+        # TODO(Mark): If we parameterised the mention scorer to score things in (0, inf)
+        # I think we could get rid of the need for this mask entirely.
+        top_span_mask = util.batched_index_select(span_mask,
+                                                  top_span_indices,
+                                                  flat_top_span_indices)
+        top_span_mention_scores = util.batched_index_select(mention_scores,
+                                                            top_span_indices,
+                                                            flat_top_span_indices)
+        top_span_starts = util.batched_index_select(span_starts,
+                                                    top_span_indices,
+                                                    flat_top_span_indices)
+        top_span_ends = util.batched_index_select(span_ends,
+                                                  top_span_indices,
+                                                  flat_top_span_indices)
+
+        # Compute indices for antecedent spans to consider.
+        max_antecedents = min(self._max_antecedents, num_spans_to_keep)
+
+        # This next section can be confusing - the returned tensors do _not_ have a batch size
```

Comment:
One more suggestion.  Here's an attempt to make this comment a little more helpful:

-----------

Now that we have our variables in terms of `num_spans_to_keep`, we need to compare span pairs to decide each span's antecedent.  Each span can only have _prior_ spans as antecedents, and we only consider up to `max_antecedents` prior spans.  So the first thing we do is construct a matrix mapping a span's index to the indices of its allowed antecedents.  Note that this is independent of the batch dimension - it's just a function of the span's position in `top_spans`.  The spans are in document order, so we can just use the relative index of the spans to know which other spans are allowed antecedents.

Once we have this matrix, we reformat our variables _again_ to get embeddings for all valid antecedents for each span.  This gives us variables with shapes like `(batch_size, num_spans_to_keep, max_antecedents, embedding_size)`, which we can use to make coreference decisions between valid span pairs.

-----------------

Along these lines, changing `antecedent_indices` to `allowed_antecedents`, `valid_antecedents`, `allowed_antecedent_indices`, or something like that, might be helpful.